Sprint5/step 1

i0ji opened the pull request:
1) Added WebSocket, reducers actions and types to serve this functionality;
2) Added middleWare functionality w/ & w/o accessToken;
3) Added new reducer and actions for WS connections;
4) Set 4 more routes;
5) Get personal orderList and common orderList depending on token availability;

- opening in a new window doesn't work yet;
- many styling elements still needed to be implemented and corrected;
- need to add Loader before we see orderList;
- profile order routing logic still needs to be implemented;
- some bugs with totalAmount of order;

i0ji committed:
feature: tooltip set

Commit:
17.03.24/b:sprint5/step-1/v:0.1.9.5
refactor: order button leads to login when unAuth;
style/feature: add style bun highlights

Commit:
feature: making order condition and highlight problem logic;

i0ji committed:
18.03.23/b:sprint5/step-1/v:0.1.9.5.1;
docs: added sprint4 readme description;
style: some small animation styling;
chore;

i0ji committed:
chore: get rid of almost all test console.logs;
fix: typing;
refactor: RootState typing; interfaces optimising; LoginPage function take out from memoise;

Commit:
v:0.1.9.5.2
refactor: changed the file structure - corresponding files moved to services;
fix: IForm small edit;
feature: custom hook file and typing;

Commit:
v:0.1.9.5.2
refactor: changed the file structure - corresponding files moved to services;
fix: IForm small edit;
feature: custom hook file and typing;

i0ji committed:
19.03.24/b:sprint5/step-1/v:0.1.9.5.3
fix: typing;
refactor: multiply type and interface changes;
feature: added navigation to '/' after logout;

i0ji committed:
style: small edits;

i0ji committed:
style: small edits;

Commit:
chore;

i0ji committed:
style: small edits;

i0ji committed:
Merge remote-tracking branch 'origin/sprint5/step-1' into sprint5/step-1

Commit:
21.03.24/b:sprint5/step-1/v:0.1.9.5.5
feature: add loaders to pages;

Commit:
21.03.24/b:sprint5/step-1/v:0.1.9.5.5
feature: add loaders to pages;
fix: long request check issue;
style: add mixins; add flex mixin; duplicate SCSS move to mixin;

Commit:
v:0.1.9.5.5

i0ji committed:
22.03.24/b:sprint5/step-1/v:0.1.9.5.5
chore: style and type small edits;

Commit:
v:0.1.9.5.5

Commit:
fix: scss fix

Commit:
style: small constructor layout edits;

Commit:
feature: added typed useSelector and useDispatch

Commit:
style: add container markup mixin; multiply small edits;
refactor: -OrderPage; +FeedPage

Commit:
feature: markup feed page

Commit:
feature: markup feed page; attempt to separate auxiliary functions into a separate file;

Commit:
style: multiply markup attempts

Commit:
style: multiply markup attempts

Commit:
chore;

Commit:
24.03.24/b:sprint5/step-1/v:0.1.9.6
feature: feed markup;

Commit:
feature: feed/number markup w/static ingredients;
style: multiply edits;

Commit:
refactor: replace <Modal> wrapper from App into Ingredients detail component;

Commit:
style: styling thumbnails; multiply style edits;

Commit:
fix: Thumbnail props type

i0ji committed:
Merge remote-tracking branch 'origin/sprint5/step-1' into sprint5/step-1

# Conflicts:
#	.yarn/install-state.gz
#	src/components/BurgerIngredients/IngredientGroup/IngredientGroupStyles.module.scss

i0ji committed:
25.03.24/b:sprint5/step-1/v:0.1.9.6.1

style: cascade attempts;

i0ji committed:
style: thumbnails set;

Commit:
v:0.1.9.6.2:
feature: history markup; new routing setting;

Commit:
v:0.1.9.6.2:
feature: history markup; new routing setting;

Commit:
style: small style edits;

Commit:
style: small style edits;

Commit:
style: small style edits;

i0ji committed:
style: small style edits;

Commit:
style: small style edits;

i0ji committed:
26.03.24/b:sprint5/step-1/v:0.1.9.6.3
global: yarn up => 4.1.1;
feature: profile/feed set;

i0ji committed:
feature: markup profile orders page; take Profile to separate folder;
style: multiply style edits; added _size for measurement vars;

i0ji committed:
feature: markup profile orders;

i0ji committed:
feature: markup profile orders;

Commit:
feature: create WS slice via configureStore; make separate folder for WS;
refactor: some types refactor;
style: minor style and code improvements; placement of variables;

i0ji committed:
feature: attempt to avoid unnecessary rerender; added

i0ji committed:
feature: attempt to avoid unnecessary rerender; added

i0ji committed:
27.03.24/b:sprint5/step-1/v:0.1.9.6.4
feature: writing and typing store, reducers and actions for WS

i0ji committed:
feature: ws attempts

Commit:
fix: rootState.ts type;
refactor: get rid of unnecessary feed store and put all logic in global store;

global: suppress unused import from vite.cfg

Commit:
feature: get WS response - message;

Commit:
feature: get WS connection; get array of orders;

i0ji committed:
28.03.24/b:sprint5/step-1/v:0.1.9.6.5;
feature: Feed markup update; get total and totalToday numbers;

i0ji committed:
feature: main feed display;
style: some minor edits;
fix: typings;

i0ji committed:
fix: order price set;

i0ji committed:
feature: order routing in progress;

Commit:
fix: details/routing

Commit:
fix: get modal out of nested components to make right implementation;

i0ji committed:
refactor: total refactor using middleware;

Commit:
refactor;

Commit:
30.03.24/b:sprint5/step-1/v:0.1.9.6.6
refactor: minor edits; explicit initial state

Commit:
refactor: add ban type disabling from ESlnit; add wsActions and second middleware; get ingredients via new middleware;
change types, actionsTypes, add second reducer/action pair;

Commit:
fix: correct data set; profile order list w/o personal filter yet;
style: some minor edits;

Commit:
31.03.24/b:sprint5/step-1/v:0.1.9.6.6
fix:delete store unused action types;
feature: routing attempts;
styles: move some common style settings to @mixin;

Commit:
31.03.24/b:sprint5/step-1/v:0.1.9.6.6
feature: personal orders check;

Commit:
feature: routing;

Commit:
feature: w/o routing;

gennady-bars commented on src/App.tsx:
```diff
@@ -14,46 +19,40 @@ import {
-  
+    const background = location.state && location.state.background;
+    const navigate = useNavigate();
     const ingredientsStatus = useSelector((state: RootState) => state.ingredients.status);
```
`(state: RootState)` уже не нужно будет типизировать вручную стейт, когда замените везде хук `useSelector`. Нужно будет удалить дублирование типизации

gennady-bars commented on src/pages/FeedPage/Feed.tsx:
```diff
@@ -0,0 +1,109 @@
+    const location = useLocation();
+
+    useEffect(() => {
+        dispatch({
+            type: wsConnect,
+            payload: `${WS_URL}/all`
+        });
```
При входе на страницу ленты заказов нужно **открывать** соединение сокета, а при покидании страницы – **закрывать**. Это делается все в одном хуке `useEffect`. То же самое и с историей заказов

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации




* Приложение ломается при входе в ленту https://disk.yandex.ru/i/atD-JY9isSOtYA  (оказалось, что после авторизации получилось войти, но ленту могут смотреть все)
* Не открывается попап в **истории**


Еще заметил ошибки:


* `(state: RootState)` уже не нужно будет типизировать вручную стейт, когда замените везде хук `useSelector`. Нужно будет удалить дублирование типизации
* Логика попапов должна быть такая же, как с ингредиентом: при вставке ссылки в браузер нужно показывать страницу с деталями
* При входе на страницу ленты заказов нужно **открывать** соединение сокета, а при покидании страницы – **закрывать**. Это делается все в одном хуке `useEffect`. То же самое и с историей заказов



--------------


А так -  код хороший

i0ji committed:
01.04.24/b:sprint5/step-1/v:0.1.9.6.7
fix: store typed issue;

i0ji committed:
feature: routing attempts;
global: add react-framer to implement smooth render;

i0ji committed:
feature: routing

i0ji committed:
feature: routing
style: remove RootState typing;

i0ji committed:
feature: routing

i0ji committed:
feature: w/o routing;

Commit:
besit uzhe =(

i0ji committed:
feature: added order request;

i0ji committed:
refactor: routing; add direct flag to OrderDetails

i0ji committed:
refactor: routing; add direct flag to OrderDetails

i0ji committed:
refactor: routing; add direct flag to OrderDetails

Commit:
02.04.24/b:sprint5/step-1/v:0.1.9.6.8
fix: route markup;common order direct and nested route set;
style: styling modal window for best behavior;

Commit:
02.04.24/b:sprint5/step-1/v:0.1.9.6.8
feature: last routing.......................................

i0ji committed:
03.04.24/b:spring5/step-1/v:0.1.9.6.7
style: profile page and orders layout refactor;
feature: routing for profile and feed pages

i0ji committed:
feature: routing set!

i0ji committed:
style: multiply style improvements and changes;
refactoring: add index for all components make refactor for it;
feature: add new alias and make refactor for it;

gennady-bars commented on src/services/utils/socketMiddleware.ts:
```diff
@@ -0,0 +1,73 @@
+
+import {refreshToken} from "utils/api.ts";
+
+export const socketMiddleware = (
+    wsActions: TwsActionTypes,
+    withTokenRefresh: boolean
+): Middleware<{}, RootState> => {
+    return (store) => {
```
Хорошо сделали `socketMiddleware`

gennady-bars commented on src/services/hooks/reduxHooks.ts:
```diff
@@ -0,0 +1,12 @@
+import {TAppThunk} from "declarations/types";
+import {RootState} from "declarations/rootState.ts";
+
+export const useSelector: TypedUseSelectorHook<RootState> = selectorHook;
+
+export const useDispatch = () => dispatchHook<TAppThunk>();
```
* Отлично типизировали хуки `useSelector` и `useDispatch`
* Отлично типизировали стейт, диспатч

gennady-bars commented on src/pages/FeedPage/Feed.tsx:
```diff
@@ -0,0 +1,125 @@
+            type: wsConnect,
+            payload: `${WS_URL}/all`
+        });
+        return (() => dispatch(wsClose()));
```
Отлично, что закрываете соединение сокетов при покидании страниц

gennady-bars commented on src/components/BurgerIngredients/IngredientGroup/IngredientGroup.tsx:
```diff
@@ -1,104 +1,104 @@
+                {ingredients.map((ingredientItem: IIngredient, i) => (
+                    <Link
+                        className={styles.ingredient_card}
+                        key={i}
```
`   key={i}`

Для атрибута `key` лучше всего использовать уникальный `id` : у карточки как раз есть такой только с нижним подчеркиванием `_id`
[Почему индекс не нужно использовать, можно прочитать тут]( https://habr.com/ru/companies/hh/articles/352150/)

gennady-bars commented on src/components/BurgerIngredients/IngredientGroup/IngredientGroup.tsx:
```diff
@@ -1,104 +1,104 @@
+
+                {/* --------------- MAPPING INGREDIENTS --------------- */}
+
+                {ingredients.map((ingredientItem: IIngredient, i) => (
```
Вам больше **не нужно типизировать** аргументы в `map`, `filter`, `reduce` и тд, если массив уже 1 раз типизирован. Тайпскрипт возьмет тип автоматически. Код станет чище

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Хорошо сделали `socketMiddleware`
* Отлично, что закрываете соединение сокетов при покидании страниц 
* Отлично типизировали хуки `useSelector` и `useDispatch`
* Отлично типизировали стейт, диспатч
* Хорошо типизировали редакс


но есть некоторые недочеты:





* В **истории** и **ленте** если **перезагрузить** страницу с открытым попапом заказа, то этот открытый попап опять должен появляться (или страница с деталями). А сейчас пустой попап  отображается после перезагрузки.  https://disk.yandex.ru/i/yfZ0BvD5rqJsJw 
* Для атрибута `key` лучше всего использовать уникальный `id` : у карточки как раз есть такой только с нижним подчеркиванием `_id`
* Вам больше **не нужно типизировать** аргументы в `map`, `filter`, `reduce` и тд, если массив уже 1 раз типизирован. Тайпскрипт возьмет тип автоматически. Код станет чище




**Можно лучше**


* Скрин https://disk.yandex.ru/i/9VpOKAVaH21ymg при открытии попапа в консоли предупреждения
* В **Опере** ломается приложение при входе в ленту (без авторизации)

i0ji committed:
04.04.24/b:sprint5/step-1/v:0.1.9.7.1
refactor: remove isDirect flag, replace it with IF condition when render OrderDetails; add _id mapping key instead of i:number in ingredients;
fix: render of OrderDetails;
feature: add error reset on successful otderSlice update;

i0ji committed:
style: some style edits;
fix: modal shadow overlaps header;
feature: add description to Loader;

i0ji committed:
v:0.1.9.7.2
refactor/feature: updated and corrected calculation of order amount and quantity of ingredients

i0ji committed:
style: fixing last img brightness; some minor style and markup edits;

Commit:
style: last thumbnail less bright;
refactor: add conditional render;

Commit:
feature: move calculating qty and price logic to separate file;
style: some markup edits for better reading;

Commit:
hotfix: replace orderData condition

Commit:
v:0.1.9.7.3
feature/style: framer option start

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

i0ji committed:
05.04.24/b:sprint5/step-1/v:0.1.9.7.3
global: Finished fifth sprint;
docs: renew README;
